### PR TITLE
Added a page about custom extension parameters

### PR DESCRIPTION
**Pull request by exyi**

### CODE REVIEW

File: Pages/control-development-binding-extension-parameters.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+```csharp
+public override JsExpression GetJsTranslation(JsExpression dataContext)
+{
+	throw new InvalidOperationException($"Can't use injected services in javascript-translated bindings.");
+}
+```
+
+
+> If you'd like to see implementation of a few more, have a look at https://github.com/riganti/dotvvm/blob/master/src/DotVVM.Framework/Compilation/ControlTree/BindingExtensionParameter.cs
```
Comment (adamjez) [inline Pages/control-development-binding-extension-parameters.md]: I would rather use markdown link, eg.: [GitHub Repository](https://github.com/riganti/dotvvm/blob/master/src/DotVVM.Framework/Compilation/ControlTree/BindingExtensionParameter.cs) 
`[GitHub Repository](https://github.com/riganti/dotvvm/blob/master/src/DotVVM.Framework/Compilation/ControlTree/BindingExtensionParameter.cs)`

File: Pages/control-development-binding-extension-parameters.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+### Registration
+
+It's very simple to register the extension parameter globally (for all pages), you simply put the modification of `DotvvmConfiguration` into a startup class:
+
+```csharp
+config.Markup.DefaultExtensionParameters.Add(
+	new InjectedServiceExtensionParameter("myCoolService", new ResolvedTypeDescriptor(typeof(MyCoolService))));
+```
+
+The parameters are part of the data context (represented by `DataContextStack` class), so they can be added by `DataContextChangeAttributes`. DotVVM compiler tracks the data context for each control and each control may modify the data context that will be inside its children or inside its templates. This is done by annotating the template property or the control by an attribute that inherits `DataContextChangeAttribute`. For example, `Repeater` changes data context of it's `ItemTemplate` into the type of element of the collection bound in `DataSource` property. This is done by annotating the property with two change attributes - the changes the type to be the type of the `DataSource` property and the second changes the type to be element of that collection:
```
Comment (adamjez) [inline Pages/control-development-binding-extension-parameters.md]: Maybe we should mention **explicitly** that you would used it when you want to register the extension parameter locally in a `DataContext`.

data context of ~~it`s~~ template &rarr; data context of **its** template

File: Pages/control-development-binding-extension-parameters.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
+## Binding Extension Parameters
+
+Any DotVVM binding is an expression that may use certain identifiers and there is a few types of them:
```
Comment (adamjez) [inline Pages/control-development-binding-extension-parameters.md]: there ~~is~~ a few types &rarr; there **are** a few types

File: Pages/control-development-binding-extension-parameters.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+* `identifier` - A name that will be used in binding expression to reference this parameter
+* `type` - Type of the parameter. If you'd do the `_index` parameter it would be `System.Int32`, the `_page` is of type `BindingPageInfo`. You can, of course, create a new type that will be used specifically for this purpose.
+* `inherit` - When the extension parameter is introduced in a specific data context (like the `_index` that is only present in Repeater-like controls), this parameter controls if the parameter will also be valid in child data contexts.
+* `GetServerEquivalent(Expression controlParameter)` - When the binding is compiled to be used in .NET code, this method is invoked to get an expression that returns the value of the parameter. You can use a reference to the current control in the code.
+* `GetJsTranslation(JsExpression dataContext)` - This get's the expression that can be used in the translated Javascript expressions. Here, you can use a reference to the knockout context to compute the value.
```
Comment (adamjez) [inline Pages/control-development-binding-extension-parameters.md]: This ~~get's~~ the &rarr; This **gets** the

File: Pages/control-development-binding-extension-parameters.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+
+For example, you may know the [`@import` directive](/Pages/advanced-ioc-di-container.md) - it basically introduces a new extension parameter that represents a service imported from `IServiceProvider`. Let's have a look at how could we implement it. First, we will need class inheriting from `BindingExtensionParameter`:
```
Comment (adamjez) [inline Pages/control-development-binding-extension-parameters.md]: how could we implement &rarr; how we could implement

**Review by adamjez — CHANGES_REQUESTED** [review]
Excellent article, good job :+1

File: Pages/control-development-binding-extension-parameters.md (hunk 1 above)
Comment (exyi) [reply Pages/control-development-binding-extension-parameters.md]: You are right that this URL is pretty long. I have added caption "have a look at the real implementation on Github"

**Review by adamjez — APPROVED** [review]

**Pull request merged**